HOUSNAV-31: Step Tracker

nicholaslyon-dt opened the pull request:
[HOUSNAV-31](https://hous-bssb.atlassian.net/browse/HOUSNAV-31)

## Overview & Purpose
Add a step tracker to the display the matches the design at all screen sizes and appropriately reflects the state of the application steps.

## Summary of Changes

- Created StepTracker component
- Created `farthestItemId` in the store to reflect the current farthest item the user is on while they may navigate back to previously answered questions
- Added logic to navigation store to figure out the state of each question in the Step Tracker display
- Updated `parseStringToComponents` to account for case where we want to show questionText without any included HTML markup
- Updated section IDs in JSON to be prefixed with a number so they stay in order when turned into an Array
- Added some icons

## Side Effects
Re-named some some CSS variables to more appropriately reflect their values when they differentiate from the imported BC Design Library values

## Testing
Navigate through 9.9.9 and see the updated Step Tracker status based on questions that are complete and skipped and sections that are complete and skipped.

## Checklist
- [x] I have written or updated vitests for this work
- [x] I have reviewed the [accessibility guidelines](https://digital.gov.bc.ca/wcag/home/) relevant to this work
- [x] I have reviewed this work with at least one screen reader (when applicable)
- [x] I have added/updated any related documentation

Commit:
HOUSNAV-31: initial step tracker functionality

laurenrevere commented:
I know this is a draft, but here are some small questions/things I noticed after taking a look:

- m^2 isn’t displayed correctly.
- Any ideas what we might want to do with P23 and P24 here? Since the question takes its value from P21, would we want to surface that in the sidebar or pick a placeholder word to use here?
Here is what is currently displayed: “Is the served by a single exit stairway or ramp?” and "Is the no more than 1.5m above adjacent ground level?"

Maybe using something like the word "feature"? Might be a question for the client.


Note: the wording of P03 will be fixed in https://hous-bssb.atlassian.net/browse/HOUSNAV-69.


Side note: the functionality is looking great so far!

Commit:
HOUSNAV-31: fixing skipping of non-component markup and answer value values

nicholaslyon-dt commented:
@laurenrevere Good catches! Updated to not skip ALL markup when skipping components (for m^2 case). Added the special case for answer value fill in as well with a default to _____. How do you think that looks/feels?

**Initial**
![Screenshot 2024-07-08 at 10 20 58 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/41478918/18f79716-28ed-4a8a-80d7-e5746621a795)

**Answered**
![Screenshot 2024-07-08 at 10 21 10 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/41478918/2e7e6ef3-a2f6-46ad-9831-22d4dc25b2c8)

Commit:
HOUSNAV-31: adding accessibility label for answer value underscore display text so something is read

Commit:
HOUSNAV-31: adding step tracker tests

Commit:
HOUSNAV-31: fixing bugs and adding tests for getAnswerValueDisplay function

laurenrevere commented:
> to not skip ALL markup when skipping components (for m^2 case). Added the special case for answer value fill in as well with a default to _____. How do you think that looks/feels?

Great updates!

Commit:
HOUSNAV-28: refactoring mobile step tracker modal to match design

wardm89 commented on packages/ui/src/variables.css:
```diff
@@ -230,6 +237,22 @@
+  --walkthrough-card-title-text-decoration-hidden: underline 1px #00000000;
+  --walkthrough-card-title-text-decoration-color-transition: underline 1px;
+  --walkthrough-card-title-text-underline-offset: 0.2em;
+  --step-tracker-heading-padding-left: var(--layout-padding-xlarge);
+  --step-tracker-padding-top: var(--layout-padding-xsmall);
+  --step-tracker-padding-left: var(--layout-padding-medium);
+  --step-tracker-padding: var(--step-tracker-padding-top)
+    var(--step-tracker-padding-left);
```
No issue just wondering if it's your intent to do this going forward rather than just using the original variables?

nicholaslyon-dt replied:
Yeah it was just because of how they had to change between mobile and desktop overall. I'm not sure I would do it this way unless I had to.

wardm89 commented on apps/web/stores/NavigationStore.ts:
```diff
@@ -143,4 +157,58 @@ export class NavigationStore {
+    );
+  };
+
+  itemWasSkipped = (
```
I think some comments on these for `itemWasSkipped`, `pastSection`, and `sectionIsComplete` would be helpful. Like a higher level "...checks if current item aka question is behind the farthest question answered; then it was skipped...". Something along those lines for each.

nicholaslyon-dt replied:
Added some comments! LMK what you think.

wardm89 reviewed (approved):
Approving since the only change requests are some comments. 

Update: Requesting changes after I just noticed the warning.

wardm89 reviewed (changes requested):
This is logging to the console after every question.
<img width="838" alt="Screenshot 2024-07-09 at 12 59 57 PM" src="https://github.com/bcgov/House-Innovations-HOUSNAV/assets/7059124/e71e2415-4c6c-43b3-a97e-23d1885b2dd9">

jbook24 commented on packages/data/json/9.10.14.json:
```diff
@@ -3,10 +3,10 @@
     "subtitle": "Volume II - 9.10.14.1-4",
     "description": "Ensure a building’s separation passes code.",
-    "startingSectionId": "intro"
+    "startingSectionId": "1_intro"
```
Why are we adding a 1_ and 2_ in front of the sectionIds?

nicholaslyon-dt replied:
This is to take advantage of the ordering of the sections. When using Object.keys to turn the object of sections into an array to iterate over for display in order, I wanted them to be in a specific order. "number-like" keys use ascending order.

jbook24 commented on apps/web/components/walkthrough/Walkthrough.tsx:
```diff
@@ -48,7 +49,6 @@ const Walkthrough = observer((): JSX.Element => {
 
   return (
     <div className="web-Walkthrough" data-testid={TESTID_WALKTHROUGH}>
```
Could we use Html5 semantics instead of nested <divs> ?

nicholaslyon-dt replied:
Good call. I hadn't paid attention to the overall structure in a while. Moved the step tracker `<section>` up into the walkthrough overall and made the `web-Walkthrough--Content` it's own `<section>`. I left this overall wrapper as a `<div>` because it's just there for styling, this component is already wrapped in a `<main>` element from the root layout and I don't think there is a better way to structure that. Thoughts?

jbook24 commented:
> This is logging to the console after every question. <img alt="Screenshot 2024-07-09 at 12 59 57 PM" width="838" src="https://private-user-images.githubusercontent.com/7059124/347049547-e71e2415-4c6c-43b3-a97e-23d1885b2dd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NDUxMTQsIm5iZiI6MTcyMDU0NDgxNCwicGF0aCI6Ii83MDU5MTI0LzM0NzA0OTU0Ny1lNzFlMjQxNS00YzZjLTQzYjMtYTk3ZS0yM2QxODg1YjJkZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDlUMTcwNjU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJjMjMwNWY4ODM2ODVkNDdjYTg0OThmM2VhNzIyN2VlY2Y5ZTM4NmJjMzA4ZWM5ZjM1OTZlMTY3MjYzMDgwNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kKZd3OQ2m0e7b_Vz5ciAMuBd37epZIyVm-yqqLZjMks">

It looks like it is happening only for P21 and in this function? 

<img width="790" alt="image" src="https://github.com/bcgov/House-Innovations-HOUSNAV/assets/126275279/a9b01ad0-a3bc-4a8d-88f9-5d0dddd91ab1">

Commit:
HOUSNAV-31: adding helper comments for logic that determines step tracker state

Commit:
HOUSNAV-31: refactored answer value retrieval to remove unneeded error messages

nicholaslyon-dt commented:
> This is logging to the console after every question. <img alt="Screenshot 2024-07-09 at 12 59 57 PM" width="838" src="https://private-user-images.githubusercontent.com/7059124/347049547-e71e2415-4c6c-43b3-a97e-23d1885b2dd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NTQ5NjMsIm5iZiI6MTcyMDU1NDY2MywicGF0aCI6Ii83MDU5MTI0LzM0NzA0OTU0Ny1lNzFlMjQxNS00YzZjLTQzYjMtYTk3ZS0yM2QxODg1YjJkZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDlUMTk1MTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTU4YTA1YWE4YzViMmUzZmRlMGUyYTkwNzkyNmZmNDcyOGQwNzdiNDY3ZWQyNjQxNjAwNDgzY2JhMjY0OWU4NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rDWKEn9BgME0mEM2NL3lZbfX7MTmtqAAOgV7S990bE8">

Ahh wasn't really a warning we needed. Refactored `getQuestionAnswerValueDisplay` and `getAnswerValueDisplay`.

Commit:
HOUSNAV-31: updating markup to be more semantic

Commit:
HOUSNAV-31: fixing rebase issue
